### ISSUE TITLE
High CPU on istio-pilot 1.20.1

### ISSUE THREAD

**Issue by ilhamderi** [istio/istio#48665]
### Is this the right place to submit this?

- [X] This is not a security vulnerability or a crashing bug
- [X] This is not a question about how to use Istio

### Bug Description

Hi team,I recently upgraded Istio from version 1.18.5 to 1.20.1 in a sequential manner. While all other Istio components appear to be functioning correctly, I'm observing unusually high CPU usage within the istiod pod.

I've carefully reviewed the istiod pod logs and haven't found any apparent errors or anomalies. The logs seem to indicate normal operation.

kubectl top po -n istio-system -l app=istiod --sort-by=cpu
`NAME                      CPU(cores)   MEMORY(bytes)   
istiod-85486b756d-w95fc   838m         133Mi`

Here that log:
`2024-01-05T07:37:08.142755Z	info	ads	ADS: new connection for node:<some-node>-6988f5664f-k6z65.staging-125
2024-01-05T07:37:08.145675Z	info	ads	CDS: PUSH request for node:<some-node>-6988f5664f-k6z65.staging resources:98 size:103.2kB cached:94/94
2024-01-05T07:37:08.147129Z	info	ads	EDS: PUSH request for node:<some-node>-6988f5664f-k6z65.staging resources:83 size:18.6kB empty:0 cached:83/83
2024-01-05T07:37:08.170082Z	info	ads	LDS: PUSH request for node:<some-node>-6988f5664f-k6z65.staging resources:76 size:324.6kB
2024-01-05T07:37:08.174314Z	info	ads	RDS: PUSH request for node:<some-node>-6988f5664f-k6z65.staging resources:48 size:41.4kB cached:48/48
2024-01-05T07:37:42.553215Z	info	ads	ADS: "172.16.0.57:53508" <some-node>-7ddf6b766c-g5mgq.staging-114 terminated
2024-01-05T07:37:42.630604Z	info	ads	ADS: new connection for node:<some-node>-7ddf6b766c-g5mgq.staging-126
2024-01-05T07:37:42.632244Z	info	ads	CDS: PUSH request for node:<some-node>-7ddf6b766c-g5mgq.staging resources:98 size:103.1kB cached:94/94
2024-01-05T07:37:42.633317Z	info	ads	EDS: PUSH request for node:<some-node>-7ddf6b766c-g5mgq.staging resources:83 size:18.6kB empty:0 cached:83/83
2024-01-05T07:37:42.643037Z	info	ads	LDS: PUSH request for node:<some-node>-7ddf6b766c-g5mgq.staging resources:76 size:324.4kB
2024-01-05T07:37:42.648369Z	info	ads	RDS: PUSH request for node:<some-node>-7ddf6b766c-g5mgq.staging resources:48 size:41.3kB cached:48/48`

istioctl analyze -A | grep -E 'Error|Warning'

`Warning [IST0108] (Pod staging/<some-node>-7ddf6b766c-g5mgq) Unknown annotation: istio.io/rev
Warning [IST0108] (Pod staging/<some-node>-6988f5664f-k6z65) Unknown annotation: istio.io/rev`

istioctl proxy-status 

`NAME                                                         CLUSTER        CDS        LDS        EDS        RDS        ISTIOD                      VERSION
<some-node>-7ddf6b766c-g5mgq.staging             Kubernetes     SYNCED     SYNCED     SYNCED     SYNCED     istiod-85486b756d-w95fc     1.20.1
<some-node>-6988f5664f-k6z65).staging             Kubernetes     SYNCED     SYNCED     SYNCED     SYNCED     istiod-85486b756d-w95fc     1.20.1
`

### Version

```Text
$ istioctl version
client version: 1.20.1
control plane version: 1.20.1
data plane version: 1.20.1 (12 proxies)

$ kubectl version
Client Version: v1.28.4
Kustomize Version: v5.0.4-0.20230601165947-6ce0bf390ce3
Server Version: v1.27.5-gke.200
```

**Comment by wulianglongrd** [comment]
Can you give the CPU profile result? You can refer to this https://github.com/istio/istio/wiki/Analyzing-Istio-Performance#cpu

**Comment by ilhamderi** [comment]
hi @wulianglongrd this is CPU profile result
![profile001](https://github.com/istio/istio/assets/19345472/7ed98bdd-0951-415d-929f-f99911e6fb86)

**Comment by wulianglongrd** [comment]
I have seen some clues and will submit an optimization PR later.

**Comment by hanxiaop** [comment]
I think there have been some updates to the in-cluster analysis. I also had an optimization that could reduce the number of analysis loops, forgot to send it.

**Comment by wulianglongrd** [comment]
> I think there have been some updates to the in-cluster analysis.

Maybe there is no performance problem in the master version?

**Comment by hanxiaop** [comment]
> > I think there have been some updates to the in-cluster analysis.
> 
> Maybe there is no performance problem in the master version?

@wulianglongrd AFAIK there should be no related changes after https://github.com/istio/istio/pull/46647, which was released in 1.20. The analysis performance is not optimal and is keeping optiming, and there were also some discussions in the past about the optimization. I've also created an issue to track dependency analysis https://github.com/istio/istio/issues/48468, which could provide a better solution to the scalability issue.

**Comment by wulianglongrd** [comment]
Got it, thanks.

**Comment by wulianglongrd** [comment]
Reopen, because @hanxiaop has other optimizations.

**Comment by hanxiaop** [comment]
I think this can be closed, the optimization will be available in 1.20.3.
